python/ray/data/_internal/table_block.py

```diff
@@ -209,9 +219,19 @@ def _zip(self, acc: BlockAccessor) -> "Block":
     def zip(self, other: "Block") -> "Block":
         acc = BlockAccessor.for_block(other)
         if not isinstance(acc, type(self)):
-            raise ValueError(
-                "Cannot zip {} with block of type {}".format(type(self), type(other))
-            )
+            if isinstance(self, TableBlockAccessor) and isinstance(
+                acc, TableBlockAccessor
+            ):
+                # If block types are different, but still both of TableBlock type, try
+                # converting both to default block type before zipping.
+                self_default, other_default = self.to_default(), acc.to_default()
+                return BlockAccessor.for_block(self_default).zip(other_default)
+            else:
+                raise ValueError(
```

Comment:
In whic case, this `ValueError` will be triggered?

Reply:
for `Block`s which do not extend `TableBlock` class, i think this will be the case. since both `ArrowBlock` and `PandasBlock` are `TableBlock`s themselves, this isn't an issue for these classes, but this would cover any case in which we have other types of `Block`s in the future.

Reply:
Got it, it's fine to keep it.